Source/WebCore/Modules/WebGPU/GPUCanvasContext.h

```diff
@@ -25,41 +25,71 @@
 
 #pragma once
 
+#include "GPUBasedCanvasRenderingContext.h"
+#include "GPUCanvasConfiguration.h"
 #include "GPUTexture.h"
 #include "HTMLCanvasElement.h"
 #include "OffscreenCanvas.h"
+#include "WebGPUSurface.h"
 #include <variant>
 #include <wtf/Ref.h>
 #include <wtf/RefCounted.h>
 #include <wtf/RefPtr.h>
 
 namespace WebCore {
 
-class GPUAdapter;
-struct GPUCanvasConfiguration;
-
-class GPUCanvasContext : public RefCounted<GPUCanvasContext> {
+class GPUCanvasContext final : public GPUBasedCanvasRenderingContext {
+    WTF_MAKE_ISO_ALLOCATED(GPUCanvasContext);
 public:
 #if ENABLE(OFFSCREEN_CANVAS)
     using CanvasType = std::variant<RefPtr<HTMLCanvasElement>, RefPtr<OffscreenCanvas>>;
 #else
     using CanvasType = std::variant<RefPtr<HTMLCanvasElement>>;
 #endif
 
-    static Ref<GPUCanvasContext> create()
-    {
-        return adoptRef(*new GPUCanvasContext());
-    }
+    static std::unique_ptr<GPUCanvasContext> create(CanvasBase&);
```

Comment:
This seems wrong, the base class (CanvasRenderingContext) is refcounted, so this should return a RefPtr, not a unique ptr

Reply:
All other canvas context creation methods return `std::unique_ptr`, hmm

Reply:
Do they subclass CanvasRenderingContext? CanvasRenderingContext has a ref() / deref() so I don't get it.

Reply:
Never mind, I looked more deeply at the code and using unique_ptr here is likely fine. CanvasRenderingContext's ref() / deref() just ref /deref its owning canvas element. Therefore, the owning canvas need to own the context via a unique_ptr. This is a special pattern.

Reply:
Right, that's what I was thinking